File: Java Projects/Coffee Machine (Easy)/Stage 5/CoffeeMachine.java
package machine;

import java.util.Scanner;

public class CoffeeMachine {
    public static void main(String[] args) {
        boolean exitNow = false;
        while (!exitNow) {
            System.out.println("Write action (buy, fill, take, remaining, exit):");
            String action = scanner.next();
            System.out.println();
            switch (action) {
                case "buy":
                    buy();
                    break;
                case "fill":
                    fill();
                    break;
                case "take":
                    take();
                    break;
                case "remaining":
                    remaining();
                    break;
                case "exit":
                    exitNow = true;
                    break;
            }
            System.out.println();
        }
    }
    
    private static void remaining() {
        String[] identifier = {
                "water",
                "milk",
                "coffee beans",
                "disposable cups",
                "money"
        };
        System.out.println("The coffee machine has:");
        for (int i = 0; i < 5; i++) {
            if (i == 4) {
                System.out.print('$');
            }
            System.out.println(inventory[i] + " of " + identifier[i]);
        }
    }

    private static void take() {
        System.out.println("I give you " + inventory[4]);
        adjustInventory(0, 0, 0, 0, -inventory[4]);
    }

    private static void fill() {
        String[] statements = {
                "Write how many ml of water do you want to add:",
                "Write how many ml of milk do you want to add:",
                "Write how many grams of coffee beans do you want to add:",
                "Write how many disposable cups of coffee do you want to add:"
        };
        for (int i = 0; i < 4; i++) {
            System.out.println(statements[i]);
            inventory[i] += scanner.nextInt();
        }
    }

    private static void buy() {
        System.out.println("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino, " +
                "back - to main menu:");
        String choice = scanner.next();
        boolean success = false;
        switch (choice) {
            case "1":
                success = buyEspresso();
                break;
            case "2":
                success = buyLatte();
                break;
            case "3":
                success = buyCappuccino();
                break;
            case "back":
                return;
        }
        if (success) {
            System.out.println("I have enough resources, making you a coffee!");
        }
    }

    private static boolean buyEspresso() {
        if (checkResources(250, 0, 16)) {
            adjustInventory(-250, 0, -16, -1, 4);
            return true;
        } else {
            return false;
        }
    }

    private static boolean buyLatte() {
        if (checkResources(350, 75, 20))
        {
            adjustInventory(-350, -75, -20, -1, 7);
            return true;
        } else {
            return false;
        }
    }

    private static boolean buyCappuccino() {
        if (checkResources(200, 100, 12)) {
            adjustInventory(-200, -100, -12, -1, 6);
            return true;
        } else {
            return false;
        }
    }

    private static boolean checkResources(int water, int milk, int beans) {
        if (water > inventory[0]) {
            System.out.println("Sorry, not enough water!");
            return false;
        } else if (milk > inventory[1]) {
            System.out.println("Sorry, not enough milk!");
            return false;
        } else if (beans > inventory[2]) {
            System.out.println("Sorry, not enough coffee beans!");
            return false;
        } else if (1 > inventory[3]) {
            System.out.println("Sorry, not enough disposable cups!");
            return false;
        }
        return true;
    }

    private static void adjustInventory(int water, int milk, int beans, int cups, int cash) {
        inventory[0] += water;
        inventory[1] += milk;
        inventory[2] += beans;
        inventory[3] += cups;
        inventory[4] += cash;
    }

    private static Scanner scanner = new Scanner(System.in);
    private static int[] inventory = {400, 540, 120, 9, 550};
}

File: Python Projects/Coffee Machine (Easy)/Stage 6/coffee_machine.py
class CoffeeMachine:
    def __init__(self):
        self.available_ml_water = 400
        self.available_ml_milk = 540
        self.available_g_beans = 120
        self.available_cups = 9
        self.available_cash = 550
        self.mode = "Menu"

    def buy(self, choice):
        if self.available_cups == 0:
            print("Sorry, not enough cups!")
            return
        if choice == "1":
            coffee_created = self.create_espresso()
        elif choice == "2":
            coffee_created = self.create_latte()
        elif choice == "3":
            coffee_created = self.create_cappuccino()
        if coffee_created:
            self.available_cups -= 1
            print("I have enough resources, making you a coffee!")

    def create_espresso(self):
        if self.available_ml_water < 250:
            print("Sorry, not enough water!")
            return False
        elif self.available_g_beans < 16:
            print("Sorry, not enough coffee beans!")
            return False
        else:
            self.available_cash += 4
            self.available_ml_water -= 250
            self.available_g_beans -= 16
            return True

    def create_latte(self):
        if self.available_ml_water < 350:
            print("Sorry, not enough water!")
            return False
        elif self.available_ml_milk < 75:
            print("Sorry, not enough milk!")
            return False
        elif self.available_g_beans < 20:
            print("Sorry, not enough coffee beans!")
            return False
        else:
            self.available_cash += 7
            self.available_ml_water -= 350
            self.available_ml_milk -= 75
            self.available_g_beans -= 20
            return True

    def create_cappuccino(self):
        if self.available_ml_water < 200:
            print("Sorry, not enough water!")
            return False
        elif self.available_ml_milk < 100:
            print("Sorry, not enough milk!")
            return False
        elif self.available_g_beans < 12:
            print("Sorry, not enough coffee beans!")
            return False
        else:
            self.available_cash += 6
            self.available_ml_water -= 200
            self.available_ml_milk -= 100
            self.available_g_beans -= 12
            return True

    def take(self):
        temp = self.available_cash
        self.available_cash = 0
        print(f"I gave you ${temp}")
        return temp

    def print_status(self):
        print("The coffee machine has: ")
        print(f"{self.available_ml_water} of water")
        print(f"{self.available_ml_milk} of milk")
        print(f"{self.available_g_beans} of coffee beans")
        print(f"{self.available_cups} of disposable cups")
        if self.available_cash > 0:
            print(f"${self.available_cash} of money")
        else:
            print("0 of money")

    def run(self, arg):
        if self.mode == "Menu":
            print("Write action (buy, fill, take, remaining, exit):")
            self.mode = "Select"
        elif self.mode == "Select":
            if arg == "buy":
                print("What do you want to buy? 1 - espresso, 2 - latte, 3 - cappuccino, back - to main menu: ")
                self.mode = "Buy"
            elif arg == "fill":
                print("Write how many ml of water do you want to add:")
                self.mode = "Fill_Water"
            elif arg == "exit":
                self.mode = "Exit"
            else:
                self.mode = "Menu"
                if arg == "take":
                    self.take()
                elif arg == "remaining":
                    self.print_status()
        elif self.mode == "Buy":
            self.mode = "Menu"
            if arg != "back":
                self.buy(arg)
        elif self.mode == "Fill_Water":
            self.available_ml_water += int(arg)
            print("Write how many ml of milk do you want to add:")
            self.mode = "Fill_Milk"
        elif self.mode == "Fill_Milk":
            self.available_ml_milk += int(arg)
            print("Write how many grams of coffee beans do you want to add:")
            self.mode = "Fill_Beans"
        elif self.mode == "Fill_Beans":
            self.available_g_beans += int(arg)
            print("Write how many disposable cups of coffee do you want to add:")
            self.mode = "Fill_Cups"
        elif self.mode == "Fill_Cups":
            self.available_cups += int(arg)
            self.mode = "Menu"


coffee_machine = CoffeeMachine()
while coffee_machine.mode != "Exit":
    if coffee_machine.mode == "Menu":
        coffee_machine.run(None)
    else:
        coffee_machine.run(input())
